sdk/src/main/java/org/coralibre/android/sdk/fakegms/nearby/ExposureNotificationClientImpl.java

```diff
@@ -1,67 +1,182 @@
 import org.coralibre.android.sdk.fakegms.nearby.exposurenotification.TemporaryExposureKey;
 import org.coralibre.android.sdk.fakegms.tasks.Task;
 import org.coralibre.android.sdk.fakegms.tasks.Tasks;
+import org.coralibre.android.sdk.internal.database.Database;
+import org.coralibre.android.sdk.internal.database.PersistentDatabase;
+import org.coralibre.android.sdk.internal.database.model.CapturedData;
+import org.coralibre.android.sdk.internal.database.model.IntervalOfCapturedData;
 
+import java.io.BufferedInputStream;
 import java.io.File;
-import java.util.Collections;
+import java.io.FileInputStream;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Set;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
 
 final class ExposureNotificationClientImpl implements ExposureNotificationClient {
+    public static final String TAG = ExposureNotificationClientImpl.class.getSimpleName();
+    private static Nearby instance = null;
+
+    private final Context context;
+    private boolean enabled = false;
+    private Database database;
+
+    private ExposureConfiguration exposureConfiguration;
+
+    ExposureNotificationClientImpl(@NonNull final Context context) {
+        this.context = context;
+    }
 
     @Override
     public Task<Void> start() {
-        // TODO: Implement!
-        return Tasks.forResult(null);
+        return Tasks.call(() -> {
+            if (!enabled) {
+                PPCP.start(context);
+                database = new PersistentDatabase(context);
+                enabled = true;
+            }
+            return null;
+        });
     }
 
     @Override
     public Task<Void> stop() {
-        // TODO: Implement!
-        return Tasks.forResult(null);
+        return Tasks.call(() -> {
+            if (enabled) {
+                PPCP.stop(context);
+                database = null;
+                enabled = false;
+            }
+            return null;
+        });
     }
 
     @Override
     public Task<Boolean> isEnabled() {
-        // TODO: Implement!
-        return Tasks.forResult(true);
+        return Tasks.forResult(enabled);
     }
 
+    /**
+     * @return the user's TEKs from the last 14 days
+     */
     @Override
     public Task<List<TemporaryExposureKey>> getTemporaryExposureKeyHistory() {
-        // TODO: Implement!
-        return Tasks.forResult(Collections.emptyList());
+        return Tasks.call(() -> {
+            Iterator<IntervalOfCapturedData> collectedPayload
+                = database.getAllCollectedPayload().iterator();
+            List<TemporaryExposureKey> result = new ArrayList<>();
+
+            while (collectedPayload.hasNext()) {
+                IntervalOfCapturedData intervalOfCapturedData = collectedPayload.next();
+                CapturedData firstCapturedData = intervalOfCapturedData.getCapturedData().get(0);
```

Comment:
Just FYI: there's need to manually iterate over an `Iterable`, you can use a for-each-loop:

```suggestion
            Iterable<IntervalOfCapturedData> collectedPayload = database.getAllCollectedPayload();
            List<TemporaryExposureKey> result = new ArrayList<>();

            for (IntervalOfCapturedData intervalOfCapturedData : collectedPayload) {
                CapturedData firstCapturedData = intervalOfCapturedData.getCapturedData().get(0);
```

Reply:
Done